packages/gasket-request/lib/request.js

```diff
@@ -0,0 +1,66 @@
+const normalizedReqMap = new WeakMap();
+
+/**
+ * @type {import('./index.js').objectFromCookieStore}
+ */
+async function objectFromCookieStore(cookieStore) {
+  return await cookieStore.getAll()
+    .reduce((acc, { name, value }) => {
+      acc[name] = value;
+      return acc;
+    }, {});
+}
+
+/**
+ * @type {import('./index.js').GasketRequest}
+ */
+export class GasketRequest {
+  static async make(requestLike) {
+    return await makeGasketRequest(requestLike);
+  }
+
+  constructor(normalizedRequest) {
+    this.headers = normalizedRequest.headers;
+    this.cookies = normalizedRequest.cookies;
+    this.query = normalizedRequest.query;
+    this.url = normalizedRequest.url;
+  }
+}
+
+/**
+ * @type {import('./index.js').makeGasketRequest}
+ */
+export async function makeGasketRequest(requestLike) {
+  if (requestLike instanceof GasketRequest) {
+    return requestLike;
+  }
+
+  const { headers, cookies = {} } = requestLike;
+
+  if (!headers) {
+    throw new Error('request argument must have headers');
+  }
+
+  let query = requestLike.query;
+
+  //
+  if (!query && 'nextUrl' in requestLike) {
+    query = requestLike.nextUrl.searchParams;
+  }
+
+  query ??= {};
+
+  if (!normalizedReqMap.has(headers)) {
```

Comment:
Not sure if this is worth addressing since it's mainly an optimization thing and duplication will probably lead to consistent results, but if you have two parallel requests awaiting a `makeGasketRequest` call they will duplicate efforts because we're not caching a Promise here.